DesKeyChecker.check() => check the change multipath is not hardened

pythcoiner opened the pull request:
This PR add a check to verify that the change descriptor is not hardened.
without this check it allow import of invalid (not  derivable change) multipath like <0;1'> into a LianaDescriptor.

closes #1183

jp1ac4 commented on src/descriptors/analysis.rs:
```diff
@@ -31,6 +31,7 @@ pub enum LianaPolicyError {
     /// The spending policy is not a valid Miniscript policy: it may for instance be malleable, or
     /// overflow some limit.
     InvalidPolicy(miniscript::Error),
+    HardenedMultipath,
```
Would it be better to call it `HardenedDerivationStep` (comparing with the comments in the function)?

jp1ac4 commented on src/descriptors/analysis.rs:
```diff
@@ -61,6 +62,7 @@ impl std::fmt::Display for LianaPolicyError {
             ),
             Self::InvalidPolicy(e) => write!(f, "Invalid Miniscript policy: {}", e),
             Self::PolicyAnalysis(e) => write!(f, "Analyzing the policy of the miniscript: {}", e),
+            LianaPolicyError::HardenedMultipath => write!(f, "Hardened multipaths are forbiden"),
```
You could use `Self::` here for consistency with other lines.

Typo: "forbidden".

jp1ac4 commented on src/descriptors/analysis.rs:
```diff
@@ -117,10 +119,12 @@ impl DescKeyChecker {
                     && der_paths.len() == 2
-                    && first_der_path.into_iter().all(|step| step.is_normal());
+                    && first_der_path.into_iter().all(|step| step.is_normal())
+                    && second_der_path.into_iter().all(|step| step.is_normal());
```
Perhaps here you could combine both checks into a single line using `der_paths.into_iter().flatten()`.

jp1ac4 reviewed (commented):
I've added a few comments.

Also, would it be possible to add a test for `check()` that produces this error?

darosior commented on src/descriptors/analysis.rs:
```diff
@@ -117,10 +119,12 @@ impl DescKeyChecker {
             if let Some(ref origin) = xpub.origin {
                 let der_paths = xpub.derivation_paths.paths();
                 let first_der_path = der_paths.first().expect("Cannot be empty");
+                let second_der_path = &der_paths[1];
```
Why would this vector always have at least two elements?

pythcoiner replied:
I guessed as `xpub` is a `MultiXPub` it should have at least 2 paths, but yeah it seems we can create a single element `DerivPaths`.

darosior reviewed (commented):
Concept ACK. I think we should consider this a bug fix and have it for v7.

pythcoiner commented:
I've dropped `LianaPolicyError::HardenedMultipath` error variant as it was not used anymore in my changes & i think `LianaPolicyError::InvalidKey` message is explicit enough and addressed @jp1ac4 comment about `der_paths`.
Also added some test cases.

jp1ac4 commented on src/descriptors/analysis.rs:
```diff
@@ -776,3 +775,72 @@ impl PartialSpendInfo {
+            "[8c3ffb6e/48'/1'/0'/2']tpubDEMt3bpQMa99W81K9h8f2FJH1C81eSd6bbSkBP8tcqQHAfSKvuGp2fz6xiVpfShzT9sKPx7DVBphChjxvNd15WcbsCca5oVz1AcUTWHxkdS/0/*";
+        let key = descriptor::DescriptorPublicKey::from_str(xpub_str).unwrap();
+        let mut checker = DescKeyChecker::new();
+        println!("{:?}", checker.check(&key));
```
nit: can this line be removed?

jp1ac4 commented on src/descriptors/analysis.rs:
```diff
@@ -776,3 +775,72 @@ impl PartialSpendInfo {
+        let mut checker = DescKeyChecker::new();
+        assert!(matches!(
+            checker.check(&key),
+            Err(LianaPolicyError::InvalidKey(_))
```
nit: I don't think you need to change it, but I was just thinking that to be sure the error contains `key`, perhaps this could be changed to:
```
assert!(matches!(
    checker.check(&key),
    Err(LianaPolicyError::InvalidKey(k)) if k == key.into()
));
```
I'm not sure if this is best practice, but it just occurred to me while reviewing.

jp1ac4 reviewed (commented):
Thanks for the updates. I just have a couple of nits, one of which I think can be ignored.

pythcoiner commented:
addressed both comments

jp1ac4 reviewed (approved):
ACK 8d495579e1f0b2bd329d3d299e707491c02a31bd.

darosior reviewed (approved):
Neat.

ACK 8d495579e1f0b2bd329d3d299e707491c02a31bd
